tasm-lib/src/list/higher_order/map.rs

```diff
@@ -463,167 +477,218 @@ mod tests {
     }
 
     #[test]
     fn test_with_raw_function_square_on_xfe_plus_another_xfe() {
-        let rawcode = RawCode::new(
-            triton_asm!(
-                square_xfe_plus_another_xfe:
-                    dup 2 dup 2 dup 2 xx_mul
-                    dup 8 dup 8 dup 8 xx_add
-                    return
-            ),
-            DataType::Xfe,
-            DataType::Xfe,
-        );
-        let snippet = Map::new(InnerFunction::RawCode(rawcode));
-        ShadowedFunction::new(snippet).test()
+        fn test_case<const N: usize>() {
+            let raw_code = InnerFunction::RawCode(RawCode::new(
+                triton_asm!(
+                    square_xfe_plus_another_xfe:
+                        dup 2 dup 2 dup 2 xx_mul
+                        dup {5 + 3 + N}
+                        dup {5 + 3 + N}
+                        dup {5 + 3 + N}
+                        xx_add
+                        return
+                ),
+                DataType::Xfe,
+                DataType::Xfe,
+            ));
+            ShadowedFunction::new(ChainMap::<N>::new(raw_code)).test();
+        }
+
+        test_case::<0>();
+        test_case::<1>();
+        test_case::<2>();
+        test_case::<3>();
+        test_case::<5>();
+        test_case::<4>();
+        test_case::<6>();
+        test_case::<7>();
     }
 
     #[test]
     fn test_u32_list_to_unit_list() {
-        let remove_elements = RawCode::new(
-            triton_asm!(remove_elements: pop 1 return),
-            DataType::U32,
-            DataType::Tuple(vec![]),
-        );
-        ShadowedFunction::new(Map::new(InnerFunction::RawCode(remove_elements))).test();
+        let f = || {
+            InnerFunction::RawCode(RawCode::new(
+                triton_asm!(remove_elements: pop 1 return),
+                DataType::U32,
+                DataType::Tuple(vec![]),
+            ))
+        };
+        test_chain_map_with_different_num_input_lists(f);
     }
 
     #[test]
     fn test_u32_list_to_u64_list() {
-        let duplicate_u32 = RawCode::new(
-            triton_asm!(duplicate_u32: dup 0 return),
-            DataType::U32,
-            DataType::U64,
-        );
-        ShadowedFunction::new(Map::new(InnerFunction::RawCode(duplicate_u32))).test();
+        let f = || {
+            InnerFunction::RawCode(RawCode::new(
+                triton_asm!(duplicate_u32: dup 0 return),
+                DataType::U32,
+                DataType::U64,
+            ))
+        };
+        test_chain_map_with_different_num_input_lists(f);
     }
 
     #[test]
     fn test_u32_list_to_u128_list_plus_x() {
-        ShadowedFunction::new(Map::new(InnerFunction::RawCode(
-            u32_to_u128_add_another_u128(),
-        )))
-        .test();
+        // this code only works with 1 input list
+        let raw_code = InnerFunction::RawCode(u32_to_u128_add_another_u128());
+        ShadowedFunction::new(Map::new(raw_code)).test();
```

Comment:
As mentioned above:
With an `init_state` method, this can be rewritten as:

```rust
    #[test]
    fn test_u32_list_to_u128_list_plus_x() {
        // this code only works with 1 input list
        let raw_code = InnerFunction::RawCode(u32_to_u128_add_another_u128());
        let snippet = Map::new(raw_code);
        let encoded_u128 = random::<u128>().encode();
        let init_state = snippet.init_state(encoded_u128, [thread_rng().gen_range(0u16..200)]);
        test_rust_equivalence_given_execution_state(
            &ShadowedFunction::new(snippet),
            init_state.into(),
        );
    }
```